File: src/lib/services/permissions.js
import { invoke } from '@tauri-apps/api/core';

/**
 * @typedef {'granted' | 'denied' | 'notdetermined' | 'restricted' | 'unknown'} PermissionState
 */

/**
 * @typedef {Object} PermissionStatus
 * @property {PermissionState} screen_recording
 * @property {PermissionState} microphone
 */

/**
 * Detect if we're running on macOS
 * @returns {boolean}
 */
function isMacOS() {
  return navigator.platform.toLowerCase().includes('mac') ||
    navigator.userAgent.toLowerCase().includes('mac');
}

/**
 * Check all permission statuses
 * @returns {Promise<PermissionStatus>}
 */
export async function checkPermissions() {
  return await invoke('check_permissions');
}

/**
 * Check screen recording permission
 * @returns {Promise<PermissionState>}
 */
export async function checkScreenRecordingPermission() {
  return await invoke('check_screen_recording_permission');
}

/**
 * Check microphone permission
 * @returns {Promise<PermissionState>}
 */
export async function checkMicrophonePermission() {
  return await invoke('check_microphone_permission');
}

/**
 * Open system settings for screen recording permission
 * @returns {Promise<void>}
 */
export async function openScreenRecordingSettings() {
  return await invoke('open_screen_recording_settings');
}

/**
 * Open system settings for microphone permission
 * @returns {Promise<void>}
 */
export async function openMicrophoneSettings() {
  return await invoke('open_microphone_settings');
}

/**
 * Open general privacy settings
 * @returns {Promise<void>}
 */
export async function openPrivacySettings() {
  return await invoke('open_privacy_settings');
}

/**
 * Test if screen capture actually works
 * @returns {Promise<boolean>}
 */
export async function testScreenCapture() {
  return await invoke('test_screen_capture');
}

/**
 * Request microphone permission via browser API
 * Returns true if granted, false otherwise
 * @returns {Promise<boolean>}
 */
export async function requestMicrophonePermission() {
  try {
    const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
    // Stop the stream immediately - we just wanted to trigger the permission
    stream.getTracks().forEach(track => track.stop());
    return true;
  } catch (e) {
    console.error('Microphone permission denied:', e);
    return false;
  }
}

/**
 * Check microphone permission via browser API
 * @returns {Promise<PermissionState>}
 */
export async function checkBrowserMicrophonePermission() {
  try {
    // @ts-ignore - permissions API
    const result = await navigator.permissions.query({ name: 'microphone' });
    switch (result.state) {
      case 'granted':
        return 'granted';
      case 'denied':
        return 'denied';
      case 'prompt':
        return 'notdetermined';
      default:
        return 'unknown';
    }
  } catch (e) {
    // Firefox doesn't support querying microphone permission
    return 'unknown';
  }
}

/**
 * Check microphone permission (cross-platform)
 * Uses native macOS plugin on macOS, falls back to browser API on other platforms
 * @returns {Promise<boolean>}
 */
export async function checkMicrophonePermissionCrossPlatform() {
  if (isMacOS()) {
    try {
      // Dynamically import the macOS plugin only on macOS
      const macPermissions = await import('tauri-plugin-macos-permissions-api');
      return await macPermissions.checkMicrophonePermission();
    } catch (e) {
      console.warn('macOS permissions plugin not available, using browser fallback:', e);
      // Fall through to browser method
    }
  }

  // Use browser API for Windows/Linux or as fallback
  const state = await checkBrowserMicrophonePermission();
  return state === 'granted';
}

/**
 * Request microphone permission (cross-platform)
 * Uses native macOS plugin on macOS, falls back to browser API on other platforms
 * @returns {Promise<boolean>}
 */
export async function requestMicrophonePermissionCrossPlatform() {
  if (isMacOS()) {
    try {
      // Dynamically import the macOS plugin only on macOS
      const macPermissions = await import('tauri-plugin-macos-permissions-api');
      return await macPermissions.requestMicrophonePermission();
    } catch (e) {
      console.warn('macOS permissions plugin not available, using browser fallback:', e);
      // Fall through to browser method
    }
  }

  // Use browser API for Windows/Linux or as fallback
  return await requestMicrophonePermission();
}

File: src-tauri/src/services/permissions.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionStatus {
    pub screen_recording: PermissionState,
    pub microphone: PermissionState,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum PermissionState {
    Granted,
    Denied,
    NotDetermined,
    Restricted,
    Unknown,
}

/// Check all permission statuses
#[tauri::command]
pub fn check_permissions() -> PermissionStatus {
    PermissionStatus {
        screen_recording: check_screen_recording_permission(),
        microphone: check_microphone_permission(),
    }
}

/// Check screen recording permission status
#[tauri::command]
pub fn check_screen_recording_permission() -> PermissionState {
    #[cfg(target_os = "macos")]
    {
        check_macos_screen_recording()
    }
    #[cfg(target_os = "windows")]
    {
        // Windows doesn't require explicit screen recording permission
        PermissionState::Granted
    }
    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
    {
        PermissionState::Unknown
    }
}

/// Check microphone permission status
#[tauri::command]
pub fn check_microphone_permission() -> PermissionState {
    // Microphone is handled by the browser's Web Speech API
    // The browser will prompt for permission automatically
    // We can't check this from Rust, so we return Unknown
    // and handle it in the frontend
    PermissionState::Unknown
}

/// Request to open system preferences for screen recording
#[tauri::command]
pub fn open_screen_recording_settings() -> Result<(), String> {
    #[cfg(target_os = "macos")]
    {
        std::process::Command::new("open")
            .arg("x-apple.systempreferences:com.apple.preference.security?Privacy_ScreenCapture")
            .spawn()
            .map_err(|e| format!("Failed to open settings: {}", e))?;
        Ok(())
    }
    #[cfg(target_os = "windows")]
    {
        // Windows doesn't have a specific screen recording settings page
        Ok(())
    }
    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
    {
        Err("Not supported on this platform".to_string())
    }
}

/// Request to open system preferences for microphone
#[tauri::command]
pub fn open_microphone_settings() -> Result<(), String> {
    #[cfg(target_os = "macos")]
    {
        std::process::Command::new("open")
            .arg("x-apple.systempreferences:com.apple.preference.security?Privacy_Microphone")
            .spawn()
            .map_err(|e| format!("Failed to open settings: {}", e))?;
        Ok(())
    }
    #[cfg(target_os = "windows")]
    {
        std::process::Command::new("ms-settings:privacy-microphone")
            .spawn()
            .map_err(|e| format!("Failed to open settings: {}", e))?;
        Ok(())
    }
    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
    {
        Err("Not supported on this platform".to_string())
    }
}

/// Open general privacy settings
#[tauri::command]
pub fn open_privacy_settings() -> Result<(), String> {
    #[cfg(target_os = "macos")]
    {
        std::process::Command::new("open")
            .arg("x-apple.systempreferences:com.apple.preference.security?Privacy")
            .spawn()
            .map_err(|e| format!("Failed to open settings: {}", e))?;
        Ok(())
    }
    #[cfg(target_os = "windows")]
    {
        std::process::Command::new("ms-settings:privacy")
            .spawn()
            .map_err(|e| format!("Failed to open settings: {}", e))?;
        Ok(())
    }
    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
    {
        Err("Not supported on this platform".to_string())
    }
}

#[cfg(target_os = "macos")]
fn check_macos_screen_recording() -> PermissionState {
    // Try to capture a tiny portion of screen to test permission
    // If it fails, permission is likely denied
    // This is a heuristic since macOS doesn't provide a direct API to check

    // We can use CGPreflightScreenCaptureAccess (macOS 10.15+)
    // but that requires linking to CoreGraphics
    // For now, we'll try a test capture

    match screenshots::Screen::all() {
        Ok(screens) => {
            if let Some(screen) = screens.first() {
                match screen.capture() {
                    Ok(_) => PermissionState::Granted,
                    Err(_) => PermissionState::Denied,
                }
            } else {
                PermissionState::Unknown
            }
        }
        Err(_) => PermissionState::Denied,
    }
}

/// Test if screen capture actually works (more reliable than permission check)
#[tauri::command]
pub fn test_screen_capture() -> Result<bool, String> {
    match screenshots::Screen::all() {
        Ok(screens) => {
            if let Some(screen) = screens.first() {
                match screen.capture() {
                    Ok(img) => {
                        // Check if we got a real image (not a blank/placeholder)
                        // A real capture should have varied pixel data
                        let data = img.to_vec();
                        if data.len() > 100 {
                            // Check if it's not all zeros or all same value
                            let first_pixel = data.get(0).unwrap_or(&0);
                            let has_variation = data.iter().take(1000).any(|p| p != first_pixel);
                            Ok(has_variation)
                        } else {
                            Ok(false)
                        }
                    }
                    Err(e) => {
                        eprintln!("Screen capture test failed: {}", e);
                        Ok(false)
                    }
                }
            } else {
                Ok(false)
            }
        }
        Err(e) => {
            eprintln!("Failed to get screens: {}", e);
            Ok(false)
        }
    }
}
